pm/queue.go

```diff
@@ -0,0 +1,148 @@
+package pm
+
+import (
+	"math/big"
+	"sync/atomic"
+)
+
+// RedeemableEmitter is an interface that describes methods for
+// emitting redeemable tickets
+type RedeemableEmitter interface {
+	// Redeemable returns a channel that a consumer can use to receive tickets that
+	// should be redeemed
+	Redeemable() chan *SignedTicket
+}
+
+// ticketQueue is a queue of winning tickets that are in line for redemption on-chain.
+// A recipient will have a ticketQueue per sender that it is actively receiving tickets from.
+// If a sender's max float is insufficient to cover the face value of a ticket it is added to the queue.
+// A ticket is pulled from the queue by the recipient when a sender has sufficient max float to cover
+// the next ticket in the queue
+//
+// Based off of: https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/queue.go
+type ticketQueue struct {
+	queue []*SignedTicket
+
+	// queueLen is an internal length counter that keeps track
+	// of the size of the queue. We maintain this counter instead
+	// of reading len(queue) in order to avoid locking the queue loop goroutine
+	queueLen int32
+
+	// in is a channel that receives queued tickets
+	in chan *SignedTicket
+
+	// availableMaxFloat is a channel that the queue uses
+	// to receive the latest available max float value
+	availableMaxFloat chan *big.Int
+
+	// redeemable is a channel that a recipient will receive
+	// redeemable tickets on as a sender's max float becomes
+	// sufficient to cover the face value of tickets
+	redeemable chan *SignedTicket
+
+	quit chan struct{}
+}
+
+func newTicketQueue() *ticketQueue {
+	return &ticketQueue{
+		in:                make(chan *SignedTicket),
+		availableMaxFloat: make(chan *big.Int),
+		redeemable:        make(chan *SignedTicket),
+		quit:              make(chan struct{}),
+	}
+}
+
+// Start initiates the main queue loop goroutine for processing tickets
+func (q *ticketQueue) Start() {
+	go q.startQueueLoop()
+}
+
+// Stop signals the ticketQueue to gracefully shutdown
+func (q *ticketQueue) Stop() {
+	close(q.quit)
+}
+
+// Add adds a ticket to the queue
+func (q *ticketQueue) Add(ticket *SignedTicket) {
+	select {
+	case q.in <- ticket:
```

Comment:
Ignore this comment; I thought there was a `default` fall through here but was mistaken. This may block if there's no listener immediately though...

Reply:
Yeah in the latest commit this would block if the queue loop is blocked while waiting on `availableMaxFloat` or blocked while sending into `redeemable`. Not that big of a problem if its called within a goroutine, but this could also be mitigated somewhat if `q.in` is a buffered channel or if we switch back to using a condvar as mentioned in other comments